File: klinara-ios/klinara-ios/Services/Files/FileModels.swift
import Foundation
import UniformTypeIdentifiers

/// Müşteri dosyaları — `apps/api/src/modules/files/dto/file.dto.ts`.
///
/// Dosya içeriği **API sürecinden geçmez**: istemci imzalı adrese doğrudan PUT
/// eder, sonra `confirm` ile kayıt açılır. Nesne anahtarı sunucuda üretilir;
/// istemciye bırakılsaydı başka bir kiracının yoluna yazmayı deneyebilirdi.

nonisolated enum FileKind: String, Codable, Sendable, CaseIterable, Identifiable {
    /// Klinik fotoğraf — sağlık verisidir (KVKK m.6), `customer.medical:*` ister.
    case photo
    /// Belge (kimlik fotokopisi, onam çıktısı) — `customer:*` yeter.
    case document

    var id: String { rawValue }

    var turkishName: String {
        switch self {
        case .photo: return "Fotoğraf"
        case .document: return "Belge"
        }
    }
}

/// Öncesi/sonrası eşlemesindeki konum.
nonisolated enum FilePosition: String, Codable, Sendable, CaseIterable, Identifiable {
    case before
    case after
    case other

    var id: String { rawValue }

    var turkishName: String {
        switch self {
        case .before: return "Öncesi"
        case .after: return "Sonrası"
        case .other: return "Diğer"
        }
    }
}

/// İndirme adresinin işaret ettiği nesne.
///
/// `thumb` yalnız ``CustomerFile/hasThumbnail`` doluyken çalışır; hazır
/// değilken sunucu `409` döner ve **tam boyuta düşmez** — ızgara farkında
/// olmadan 25 MB'lık nesneler indirirdi.
nonisolated enum FileVariant: String, Sendable {
    case original
    case thumb
}

/// Sunucunun kabul ettiği içerik tipleri. Beyaz liste, kara liste değil;
/// `image/svg+xml` bilinçli olarak YOK — SVG çalıştırılabilir içerik taşır.
nonisolated enum FileContentType {
    static let allowed = [
        "image/jpeg",
        "image/png",
        "image/webp",
        "image/heic",
        "application/pdf",
    ]

    /// Sunucudaki `UPLOAD_MAX_BYTES` varsayılanı (25 MB). Yükleme öncesi
    /// küçültme hedefi; son söz yine sunucunun.
    static let maxBytes = 25 * 1024 * 1024

    /// Dosya seçicinin göstereceği tipler — ``allowed`` listesinden türetiliyor.
    ///
    /// Ayrı bir liste yazılsaydı, sunucunun kabul ettiği bir tip seçicide
    /// görünmeyebilir ya da tersi olur, kullanıcı seçtikten SONRA reddedilirdi.
    static let allowedUTTypes: [UTType] = allowed.compactMap { UTType(mimeType: $0) }

    /// Gövdenin gerçek içerik tipi.
    ///
    /// Sıra bilinçli: önce **baytların imzası**, sonra dosya uzantısı. Uzantı
    /// bir iddiadır (`.pdf` uzantılı bir JPEG mümkün), imza ise içeriğin
    /// kendisidir; `presign` beyanı ile nesnenin gerçeği ayrışırsa imzalı PUT
    /// `Content-Type` uyuşmazlığına düşer.
    ///
    /// Tanınmayan içerik `nil` döner — **varsayılan tip yok**. Bir varsayılan,
    /// yanlış tipi sessizce kaydetmenin ta kendisidir.
    static func detect(data: Data, filenameExtension: String? = nil) -> String? {
        if let signature = signature(of: data) { return signature }
        guard let ext = filenameExtension, !ext.isEmpty,
              let type = UTType(filenameExtension: ext),
              let mime = type.preferredMIMEType,
              allowed.contains(mime)
        else { return nil }
        return mime
    }

    /// Sihirli bayt dizileri. WebP ve HEIC kapsayıcı biçimler: ilk dört bayt
    /// tipi vermiyor, marka alanına bakmak gerekiyor.
    private static func signature(of data: Data) -> String? {
        func matches(_ bytes: [UInt8], at offset: Int) -> Bool {
            guard data.count >= offset + bytes.count else { return false }
            let start = data.index(data.startIndex, offsetBy: offset)
            return Array(data[start..<data.index(start, offsetBy: bytes.count)]) == bytes
        }

        if matches(Array("%PDF".utf8), at: 0) { return "application/pdf" }
        if matches([0xFF, 0xD8, 0xFF], at: 0) { return "image/jpeg" }
        if matches([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A], at: 0) { return "image/png" }
        if matches(Array("RIFF".utf8), at: 0), matches(Array("WEBP".utf8), at: 8) {
            return "image/webp"
        }
        if matches(Array("ftyp".utf8), at: 4) {
            let brands = ["heic", "heix", "hevc", "heim", "heis", "hevm", "hevs", "mif1", "msf1"]
            if brands.contains(where: { matches(Array($0.utf8), at: 8) }) { return "image/heic" }
        }
        return nil
    }

    /// Listede gösterilecek ad. Ham MIME kullanıcıya bir şey anlatmıyor.
    static func turkishName(of mimeType: String) -> String {
        switch mimeType {
        case "application/pdf": return "PDF belge"
        case "image/jpeg": return "JPEG görsel"
        case "image/png": return "PNG görsel"
        case "image/webp": return "WebP görsel"
        case "image/heic": return "HEIC görsel"
        default: return "Belge"
        }
    }
}

nonisolated struct CustomerFile: Codable, Sendable, Identifiable, Equatable {
    let id: String
    let customerId: String
    let groupId: String?
    let kind: FileKind
    let position: FilePosition
    let mimeType: String
    let sizeBytes: Int
    let sha256: String?
    /// Küçük görsel hazır mı — kuyruk işi tamamlanınca dolar.
    let hasThumbnail: Bool
    let takenAt: Date?
    let uploadedBy: String?
    let createdAt: Date
}

nonisolated struct CustomerFileGroup: Codable, Sendable, Identifiable, Equatable {
    let id: String
    let title: String
    let bodyArea: String?
    let serviceId: String?
    let files: [CustomerFile]
    let createdAt: Date

    func file(at position: FilePosition) -> CustomerFile? {
        files.first { $0.position == position }
    }
}

// MARK: - İstekler

nonisolated struct PresignUploadInput: Encodable, Sendable, Equatable {
    let customerId: String
    let contentType: String
    /// Bayt. Sunucu üst sınırı aşarsa reddeder ve `confirm` adımında boyutu
    /// **nesnenin kendisinden** okur — bu değer bir beyandır.
    let sizeBytes: Int
    let kind: FileKind
}

nonisolated struct PresignUploadResponse: Decodable, Sendable, Equatable {
    /// `confirm` adımına aynen verilecek anahtar.
    let storageKey: String
    /// İstemcinin doğrudan PUT edeceği adres.
    let uploadUrl: String
    /// PUT isteğinde AYNEN gönderilmesi gereken `Content-Type`.
    let contentType: String
    let expiresAt: Date
}

nonisolated struct ConfirmFileInput: Encodable, Sendable, Equatable {
    let storageKey: String
    let kind: FileKind
    var position: FilePosition?
    var groupId: String?
    /// İçeriğin sha256 özeti (hex). Worker nesneyi indirirken doğruluyor;
    /// uyuşmazlıkta kayıt `pending`e düşüyor.
    var sha256: String?
    var takenAt: Date?
}

nonisolated struct CreateFileGroupInput: Encodable, Sendable, Equatable {
    let title: String
    var bodyArea: String?
    var serviceId: String?
}

nonisolated struct DownloadURL: Decodable, Sendable, Equatable {
    let url: String
    let expiresAt: Date
}

File: klinara-android/app/src/main/java/com/klinara/android/services/files/FileModels.kt
package com.klinara.android.services.files

import com.klinara.android.services.networking.InstantSerializer
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import java.time.Instant

/**
 * Dosya türü.
 *
 * **[Photo] sağlık verisidir** (KVKK m.6) ve `customer.medical:*` iznine bağlıdır;
 * [Document] (kimlik fotokopisi, sözleşme) `customer:*` iznine. Ayrım sunucuda
 * `customer_files.kind` üzerinden yapılıyor ve istemci de aynı ayrımı çizmek zorunda —
 * bir klinik fotoğrafı belge gibi göstermek, izin kapısını delmek olurdu.
 */
@Serializable(with = FileKindSerializer::class)
enum class FileKind(val wire: String) {
    Photo("photo"),
    Document("document"),
    Unknown("unknown"),
    ;

    val turkishName: String
        get() =
            when (this) {
                Photo -> "Fotoğraf"
                Document -> "Belge"
                Unknown -> "Bilinmeyen"
            }

    companion object {
        fun from(wire: String): FileKind = entries.firstOrNull { it.wire == wire } ?: Unknown
    }
}

internal object FileKindSerializer : KSerializer<FileKind> {
    override val descriptor = PrimitiveSerialDescriptor("FileKind", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): FileKind = FileKind.from(decoder.decodeString())

    override fun serialize(
        encoder: Encoder,
        value: FileKind,
    ) = encoder.encodeString(value.wire)
}

/** Öncesi/sonrası konumu. */
@Serializable(with = FilePositionSerializer::class)
enum class FilePosition(val wire: String) {
    Before("before"),
    After("after"),
    Other("other"),
    ;

    val turkishName: String
        get() =
            when (this) {
                Before -> "Öncesi"
                After -> "Sonrası"
                Other -> "Diğer"
            }

    companion object {
        fun from(wire: String): FilePosition = entries.firstOrNull { it.wire == wire } ?: Other
    }
}

internal object FilePositionSerializer : KSerializer<FilePosition> {
    override val descriptor = PrimitiveSerialDescriptor("FilePosition", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): FilePosition = FilePosition.from(decoder.decodeString())

    override fun serialize(
        encoder: Encoder,
        value: FilePosition,
    ) = encoder.encodeString(value.wire)
}

/**
 * İndirilecek nesnenin hangi hâli.
 *
 * ⚠️ [Thumb] **yalnız fotoğraflarda ve küçük görsel üretildikten SONRA** anlamlı.
 * Hazır değilken sunucu `409` döner ve istemci **tam boyuta DÜŞMEZ**: sessiz düşüş,
 * 30 fotoğraflı bir ızgaranın farkında olmadan 25 MB'lık nesneler indirmesi demekti.
 */
enum class FileVariant(val wire: String) {
    Original("original"),
    Thumb("thumb"),
}

/**
 * İzin verilen içerik tipleri — **beyaz liste, kara liste değil**.
 *
 * `image/svg+xml` bilinçle YOK: SVG çalıştırılabilir içerik taşır. Yarın eklenecek bir
 * tipin sessizce geçmesi yerine açıkça eklenmesi gerekiyor.
 */
object FileContentType {
    const val JPEG = "image/jpeg"
    const val PNG = "image/png"
    const val WEBP = "image/webp"
    const val HEIC = "image/heic"
    const val PDF = "application/pdf"

    val allowed: List<String> = listOf(JPEG, PNG, WEBP, HEIC, PDF)

    /** Sunucunun `UPLOAD_MAX_BYTES` sınırı. */
    const val MAX_BYTES: Long = 25L * 1024 * 1024

    fun isAllowed(contentType: String): Boolean = contentType in allowed

    fun turkishName(contentType: String): String =
        when (contentType) {
            JPEG -> "JPEG görsel"
            PNG -> "PNG görsel"
            WEBP -> "WebP görsel"
            HEIC -> "HEIC görsel"
            PDF -> "PDF belge"
            else -> "Bilinmeyen tür"
        }

    /**
     * İçerik tipini **SİHİRLİ BAYTTAN** tespit eder; uzantı yalnız ikinci sırada.
     *
     * Uzantıya güvenmek, `.jpg` diye adlandırılmış bir SVG'yi kabul etmek olurdu.
     * Bilinmeyen içerik `null` döner ve **varsayılan tip ATANMAZ**: tanımadığımız bir
     * dosyayı JPEG sanıp yüklemek, sunucunun reddedeceği bir nesneyi depoya yazmaktı.
     */
    fun detect(
        data: ByteArray,
        fileExtension: String? = null,
    ): String? {
        fromMagicBytes(data)?.let { return it }
        return fileExtension?.lowercase()?.let { ext ->
            when (ext) {
                "jpg", "jpeg" -> JPEG
                "png" -> PNG
                "webp" -> WEBP
                "heic", "heif" -> HEIC
                "pdf" -> PDF
                else -> null
            }
        }
    }

    private fun fromMagicBytes(data: ByteArray): String? {
        if (data.size < MIN_SNIFF_BYTES) return null

        if (data.startsWith(PDF_MAGIC)) return PDF
        if (data.startsWith(JPEG_MAGIC)) return JPEG
        if (data.startsWith(PNG_MAGIC)) return PNG
        // RIFF....WEBP — imza iki parçalı, 0 ve 8. offsetlerde.
        if (data.startsWith("RIFF".toByteArray()) && data.matchesAt(RIFF_TYPE_OFFSET, "WEBP".toByteArray())) {
            return WEBP
        }
        // ftyp + HEIC markası, 4. ve 8. offsetlerde.
        if (data.matchesAt(FTYP_OFFSET, "ftyp".toByteArray())) {
            val brand = String(data, HEIC_BRAND_OFFSET, HEIC_BRAND_LENGTH)
            if (brand in HEIC_BRANDS) return HEIC
        }
        return null
    }

    private fun ByteArray.startsWith(prefix: ByteArray): Boolean = matchesAt(0, prefix)

    private fun ByteArray.matchesAt(
        offset: Int,
        expected: ByteArray,
    ): Boolean {
        if (size < offset + expected.size) return false
        return expected.indices.all { this[offset + it] == expected[it] }
    }

    /** İmzalar — hepsi spesifikasyondan; sihirli sayı değil, TANIM. */
    private val PDF_MAGIC = "%PDF".toByteArray()
    private val JPEG_MAGIC = hex("FFD8FF")
    private val PNG_MAGIC = hex("89504E470D0A1A0A")

    /** Onaltılık imzayı bayta çevirir — spesifikasyondaki gösterimle aynı okunsun diye. */
    private fun hex(value: String): ByteArray =
        value.chunked(2).map { it.toInt(HEX_RADIX).toByte() }.toByteArray()

    private const val HEX_RADIX = 16

    private const val MIN_SNIFF_BYTES = 12
    private const val RIFF_TYPE_OFFSET = 8
    private const val FTYP_OFFSET = 4
    private const val HEIC_BRAND_OFFSET = 8
    private const val HEIC_BRAND_LENGTH = 4
    private val HEIC_BRANDS = setOf("heic", "heix", "hevc", "heim", "heis", "mif1", "msf1")
}

/** Müşteri dosyası. */
@Serializable
data class CustomerFile(
    val id: String,
    val customerId: String,
    val groupId: String? = null,
    val kind: FileKind = FileKind.Document,
    val position: FilePosition = FilePosition.Other,
    val mimeType: String = "",
    val sizeBytes: Long = 0,
    val sha256: String? = null,
    /** Küçük görsel HAZIR mı — kuyruk işi tamamlanınca dolar. */
    val hasThumbnail: Boolean = false,
    @Serializable(with = InstantSerializer::class)
    val takenAt: Instant? = null,
    val uploadedBy: String? = null,
    @Serializable(with = InstantSerializer::class)
    val createdAt: Instant? = null,
)

/** Öncesi/sonrası grubu. */
@Serializable
data class FileGroup(
    val id: String,
    val title: String,
    val bodyArea: String? = null,
    val serviceId: String? = null,
    val files: List<CustomerFile> = emptyList(),
    @Serializable(with = InstantSerializer::class)
    val createdAt: Instant? = null,
) {
    fun file(at: FilePosition): CustomerFile? = files.firstOrNull { it.position == at }
}

/** `POST uploads/presign` yanıtı. */
@Serializable
data class PresignUpload(
    val storageKey: String,
    val uploadUrl: String,
    /** PUT isteğinde AYNEN gönderilmeli — imza bunu kapsıyor. */
    val contentType: String,
    @Serializable(with = InstantSerializer::class)
    val expiresAt: Instant? = null,
)

/** `GET files/:id/download-url` yanıtı. */
@Serializable
data class DownloadUrl(
    val url: String,
    @Serializable(with = InstantSerializer::class)
    val expiresAt: Instant? = null,
)
